Sources/Flows/OAuth2.swift

```diff
@@ -366,8 +366,12 @@ open class OAuth2: OAuth2Base {
 				do {
 					let data = try response.responseData()
 					let json = try self.parseRefreshTokenResponseData(data)
-					if response.response.statusCode >= 400 {
-						self.clientConfig.refreshToken = nil
+					switch response.response.statusCode {
+					case 500:
```

Comment:
Should we not return serverError for any 5xx status code?

Reply:
the comment in the error class said it only handled 500. 

```
/// A 500 was thrown.
 case serverError
```

Since it is used in other places as well, it would require migration for all usages. That's why I left it like this.

Reply:
It's fine to not delete the token here, the `didFail` method will be called with an error and the library user can then take action accordingly.

Indeed, `serverError` is used for the specific `server_error` OAuth2 error code. How about a new `serverErrorWithStatus(Int)` error that can be thrown here? If so we could adapt to `clientErrorWithStatus(Int)`.

Reply:
Also, don't forget to add your name to CONTRIBUTORS!

Reply:
I looked into it a bit more and figured out that in case the response was not parsed for the JSON, it would return `.invalidGrant` error. I changed the logic a bit to only try and parse the response if the status code was 2xx. Otherwise, the other errors would not have been thrown. It makes sense to only try to parse the response if the request returned successfully.

Reply:
The JSON should always be parsed, in the case of 400+ it contains error information.

Also, please undo the changes to `Keychain.swift`.

Reply:
From my test, for a 400, the json parsing fails and returns `invalidGrant`. which means the new errors will never be thrown. In case the parsing fails and it caught by the catch block then the new errors make no sense - unless the backend returns a parsable json but a status code different than 200, which is very unlikely

Reply:
In a correctly behaving server, any 400 response status is accompanied by valid JSON with an `error` key and optional an `error_description` key, see here: https://tools.ietf.org/html/rfc6749#section-5.2

Reply:
Yes, in this case, the parsing of the json would already throw the right error, like `.invalidGrant`. For the other errors, assuming they happen, like 500 for example, where there will be no json response, the response code will not be handled through the new code, but through the parsing, which will fail returning some of the predefined errors.
Considering the 4xx errors are handled by the json parsing according to the standard, I would remove them and only leave the 5xx, `.serverErrorWithStatus` error case and unify the 4xx and 2xx cases in the switch, since the same code can handle both of them

Reply:
Ah I misunderstood, I thought you said that any `400` will always throw `invalidGrant`.

So in a correctly working server, a 400 will never even touch the `response.response.statusCode` switch. I forgot about that. It seems we should keep the 400 case in case the server returns gibberish and throw a new `clientErrorWithStatus(Int)` error. And do the same for 500 with `serverErrorWithStatus(Int)`. Do you think that makes sense?

Reply:
I spend quite some time thinking what would make most sense. I merged the cases where te json contains valuable information - 4xx and 2xx and left the serverError handling separate. I removed the `clientErrorWithStatus(Int)` because I do not believe it makes sense since we will never return it, but always return the result of the json parsing.

Let me know what you think.